### PR TITLE
Replace download package with progressive-downloader, resolves #1378

### LINKED ISSUE #1322: Proxy support
Shortly after starting up 0.14.8beta version of the installer, I get this message:

installer-error

The logs say:

```
$ ./ubports-installer.0.4.18-beta.AppImage -D -V
info: Welcome to the UBports Installer version 0.4.18-beta!
debug: killing all running adb servers
command: adb: {"args":["-P",5037,"kill-server"],"error":null,"stdout":"","stderr":""}
debug: starting adb server on port 5037
command: adb: {"args":["-P",5037,"start-server"],"error":null,"stdout":"","stderr":"* daemon not running; starting now at tcp:5037\n* daemon started successfully\n"}
command: adb: {"args":["-P",5037,"shell","echo","."],"error":null,"stdout":".\r\n","stderr":""}
command: adb: {"args":["-P",5037,"shell","getprop","ro.product.device"],"error":null,"stdout":"turbo\r\n","stderr":""}
error: getDeviceSelects error: Error: timeout of 7500ms exceeded
error: getDeviceName error: Error: timeout of 7500ms exceeded
```

Version 0.14.4beta works fine (and of course, my internet works fine too).

### LINKED ISSUE #1376: Implement heimdall:flash step
To support samsung devices, we need `heimdall:*` steps analogous with their equivalent operations in `fastboot`. The most important one:

```
     {
          "type": "heimdall:flash",
          "flash": [
            {
              "partition": "recovery",
              "file": "recovery-verified.img",
              "group": "firmware"
            },
            {
              "partition": "boot",
              "file": "recovery-verified.img",
              "group": "firmware"
            }
          ]
        }
```

### LINKED ISSUE #1378: Replace download package with a well-maintained library
The `download` package is kind of a hacky abomination. While one might argue that this fits in great with the rest of the installer, it causes annoying issues every now and again.

### PR DESCRIPTION
**Pull request by NeoTheThird**
Replaces the abomination that is `download` with a [fancy solution](https://www.npmjs.com/package/progressive-downloader), resolves #1378. Also includes Heimdall, resolves #1376!

Minor other changes:

- I also found some remnants of `request`, that were still in here for some reason. Nuked with great prejudice.
- Pastes now expire after one year, so we stop flooding paste.ubuntu.com as much.
- Added support for hashing algorithms other than sha256.
- Got rid a lot of unneeded dependencies.
- Resolves #1322 if merged with https://github.com/ubports/system-image-node-module/pull/24, since axios supports the `http_proxy`/`https_proxy` env vars (https://www.golinuxcloud.com/set-up-proxy-http-proxy-environment-variable/).
- Minor code cleanup
- Re-structure generated bug reports
- Cleaner cache directory structure
- Improved re-usability of downloaded files while ensuring everything is up-to-date

### CODE REVIEW

**Comment by NeoTheThird** [comment]
Currently blocked by #1374, since we require changes from there and i'm in no mood for cherry-picking. I'll also take a swing at #1322, since that's annoying as hell, and add the new lib to system-image-node-module. *Stand by while developers are developing*.

**Comment by Flohack74** [comment]
Uh this ia a big one, I think we need to do interactive review on that.

**Comment by NeoTheThird** [comment]
might be a good idea :)

File: src/devices.js [hunk 1]
```diff
@@ -261,6 +286,18 @@ function installStep(step) {
         global.mainEvent.emit("user:write:under", "Resuming boot");
         fastboot.continue();
       };
+    case "heimdall:flash":
+      return () => {
+        global.mainEvent.emit("user:write:working", "particles");
+        global.mainEvent.emit("user:write:status", "Flashing firmware", true);
+        global.mainEvent.emit(
+          "user:write:under",
+          "Flashing firmware partitions using fastboot"
```
[the remark this answers is not in the document]
Comment (NeoTheThird) [reply src/devices.js]: good catch :)

**Review by Flohack74 — APPROVED** [review]

**Pull request merged**